src/System.Private.CoreLib/shared/System/IO/MemoryStream.cs

```diff
@@ -226,18 +226,26 @@ internal int InternalGetPosition()
             return _position;
         }
 
-        // PERF: Takes out Int32 as fast as possible
-        internal int InternalReadInt32()
+        // PERF: Takes out ReadOnlySpan as fast as possible
+        internal ReadOnlySpan<byte> InternalReadSpan(int count)
         {
             EnsureNotClosed();
 
-            int pos = (_position += 4); // use temp to avoid a race condition
-            if (pos > _length)
+            int origPos = _position;
+            int newPos = origPos + count;
+            
+            // Check validity of what we got
+            if (newPos < 0)
+                throw new IOException(SR.IO_StreamTooLong);
```

Comment:
This is not the right error message to throw here. If `newPos` is negative here, it means that the read was beyond the end of the stream.

It would be better to change the following check to:

`if ((uint)newPos > (uint)_length)`

It would cover the integer overflow case without any additional overhead.

Reply:
fixed.